### PR TITLE
AIML skill

### PR DESCRIPTION
**Pull request by acriptis**
Implement AIML Skill adapter for DeepPavlov Agent

### CODE REVIEW

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    def process_step(self, utterance_str: str, user_id: any):
+        response = self.kernel.respond(utterance_str, sessionID=user_id)
+        # here put your estimation of confidence:
+        confidence = self.default_confidence
+        if response:
+            # print(f"AIML responds: {response}")
+            pass
+        else:
+            # print("AIML responses silently...")
+            response = self.null_response
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: Are you sure that the skill should have the same confidence for both types of responses?

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    def _generate_user_id(self):
+        """
+        Here you put user id generative logic if you want to implement it in the skill.
+
+        Although it is better to delegate user_id generation to Agent Layer
+        Returns: int
+
+        """
+        import datetime
+        return datetime.datetime.utcnow().timestamp() * 10e6
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: No need for local import. You can also use [uuid module](https://docs.python.org/3.6/library/uuid.html) for id generation.

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        confidence = self.default_confidence
+        if response:
+            # print(f"AIML responds: {response}")
+            pass
+        else:
+            # print("AIML responses silently...")
+            response = self.null_response
+        return response, confidence
+
+    def _generate_user_id(self):
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: Return types here and everywhere

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        """
+        # grasp user_ids from states batch.
+        # We expect that skill receives None or dict of state for each utterance.
+        # if state has user_id then skill uses it, otherwise it generates user_id and calls the
+        # user with this name in further.
+
+        # In this implementation we use current datetime for generating uniqe ids
+        output_states_batch = []
+        user_ids = []
+        for idx, each_state in enumerate(states_batch):
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: `idx` is not used

File: deeppavlov/skills/aiml_skill/requirements.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+python-aiml==0.9.1
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/requirements.txt]: This bothers me. When installing `deeppavlov` from `pip` user will not see this requirement.
I see a solution in creating a deeppavlov config for the aiml skill that will have a link on this requirement that will be moved to `/deeppavlov/requirements`

File: deeppavlov/skills/aiml_skill/test_aiml_skill.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+import unittest
+import os
+
+from deeppavlov.agents.default_agent.default_agent import DefaultAgent
+from deeppavlov.agents.processors.highest_confidence_selector import HighestConfidenceSelector
+from deeppavlov.skills.aiml_skill.aiml_skill import AIMLSkill
+
+
+class TestAIMLSkill(unittest.TestCase):
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/test_aiml_skill.py]: We use `pytest`. Tests can be found in `/tests/`

File: deeppavlov/skills/aiml_skill/README.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+This skill wraps python-aiml library and allows developer to integrate AIML scripts into DeepPavlov dialog system.
+
+Try: `test_aiml_skill.py` to understand usage.
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/README.md]: Please add usage examples.

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
+import random
+import re
+import os
+from typing import List, Tuple, Optional, Union
+from deeppavlov.core.skill.skill import Skill
+import aiml
```
[the remark this answers is not in the document]
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Fixed

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    def _load_scripts(self):
+        """
+        Scripts are loaded recursively from files with extensions .xml and .aiml
+        Returns:
+
+        """
+        # learn kernel to all aimls in directory tree:
+
+        for root, dirs, files in os.walk(self.path_to_aiml_scripts):
```
[the remark this answers is not in the document]
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Fixed

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        Args:
+            path_to_aiml_scripts: string path to folder with AIML scripts
+            null_response: Response string to answer if no AIML Patterns matched
+            default_confidence: The default confidence.
+        """
+
+        if not path_to_aiml_scripts:
+            import os
+            cur_dir = os.path.dirname(os.path.abspath(__file__))
+            path_to_aiml_scripts = cur_dir + "/aiml_scripts"
```
[the remark this answers is not in the document]
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Fixed

File: deeppavlov/skills/aiml_skill/aiml_skill.py (hunk 1 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: In updated code I've added attribute `null_confidence` with lower value and set lower confidence if no patterns matched in AIML scripts

File: deeppavlov/skills/aiml_skill/aiml_skill.py (hunk 2 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Fixed

File: deeppavlov/skills/aiml_skill/aiml_skill.py (hunk 3 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Fixed

File: deeppavlov/skills/aiml_skill/aiml_skill.py (hunk 4 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: fixed

File: deeppavlov/skills/aiml_skill/test_aiml_skill.py (hunk 6 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/test_aiml_skill.py]: fixed

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    def _load_scripts(self) -> None:
+        """
+        Scripts are loaded recursively from files with extensions .xml and .aiml
+        Returns: None
+
+        """
+        # learn kernel to all aimls in directory tree:
+        all_files = sorted(self.path_to_aiml_scripts.rglob('*.*'))
+        for each_file_path in all_files:
+            if each_file_path.suffix in ['.aiml', '.xml']:
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: There is no error nor a warning if there are no scripts in `self.path_to_aiml_scripts`

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
+from pathlib import Path
+from typing import Union
+from typing import Tuple, Optional, List
+from logging import getLogger
+import aiml
+import uuid
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: PEP 8 imports order

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            path_to_aiml_scripts: string path to folder with AIML scripts
+            null_response: Response string to answer if no AIML Patterns matched
+            positive_confidence: The confidence of response if response was found in AIML scripts
+            null_confidence: The confidence when AIML scripts has no rule for responding and system returns null_response
+        """
+
+        if not path_to_aiml_scripts:
+            cur_dir = Path(__file__).absolute().parent
+            self.path_to_aiml_scripts = cur_dir.joinpath("aiml_scripts")
+            log.warning(f"path_to_aiml_scripts is not provided, use default path: `{self.path_to_aiml_scripts}`")
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: Doesn't apply anymore. `path_to_aiml_scripts` should be a required argument.

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+            response = self.null_response
+            confidence = self.null_confidence
+        return response, confidence
+
+    def _generate_user_id(self) -> str:
+        """
+        Here you put user id generative logic if you want to implement it in the skill.
+
+        Although it is better to delegate user_id generation to Agent Layer
+        Returns: int
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: Is it `str` or `int`?

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+        """
+        return uuid.uuid1()
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]:
```suggestion
        return uuid.uuid1().hex
```
for `str` result

File: tests/test_aiml_skill.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
+import pytest
+from pathlib import Path
```
Comment (yoptar) [inline tests/test_aiml_skill.py]: PEP 8 imports order

File: tests/test_aiml_skill.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+from deeppavlov.agents.processors.highest_confidence_selector import HighestConfidenceSelector
+from deeppavlov.skills.aiml_skill.aiml_skill import AIMLSkill
+
+
+class TestAIMLSkill:
+    def setup(self):
+        print("basic setup into class")
+        cur_dir = Path(__file__).absolute().parent.parent
+
+        path_to_aiml_scripts = cur_dir.joinpath("deeppavlov", "skills","aiml_skill", "aiml_scripts")
```
Comment (yoptar) [inline tests/test_aiml_skill.py]: Does not apply anymore

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 18]
```diff
@@ -0,0 +1,139 @@
+from pathlib import Path
+from typing import Union
+from typing import Tuple, Optional, List
+from logging import getLogger
+import aiml
+import uuid
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.skill.skill import Skill
+log = getLogger(__name__)
```
Comment (my-master) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: empty line

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+        self.null_confidence = null_confidence
+        self.null_response = null_response
+        self.kernel = aiml.Kernel()
+        # to block AIML output:
+        self.kernel._verboseMode = False
+        self._load_scripts()
+
+    def _load_scripts(self) -> None:
+        """
+        Scripts are loaded recursively from files with extensions .xml and .aiml
```
Comment (my-master) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: I understand this just a wrapper, but would be nice to have an example of these .xml or .aiml files structure in the docstring.

Comment (yoptar) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Or maybe a link to them?

File: deeppavlov/skills/aiml_skill/aiml_skill.py (hunk 14 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Really it is uuid.UUID object =)
Fixed docs

File: deeppavlov/skills/aiml_skill/aiml_skill.py (hunk 15 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: It works with uuid.UUID is it better to make string result?

File: tests/test_aiml_skill.py (hunk 16 above)
Comment (acriptis) [reply tests/test_aiml_skill.py]: Hm, don't catch idea.
```Imports should be grouped in the following order:

Standard library imports.
Related third party imports.
Local application/library specific imports.
You should put a blank line between each group of imports.
```
Is this comment related to order of imports or for lack of blank lines?

Comment (acriptis) [reply tests/test_aiml_skill.py]:
```
from pathlib import Path
import pytest
```

Is this right order?

Comment (yoptar) [reply tests/test_aiml_skill.py]: `pytest` is not in a standard library

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+    AIML uses directory with AIML scripts which are loaded at initialization and used as patterns
+    for answering at each step.
+    """
+
+    def __init__(self,
+                 path_to_aiml_scripts: Union[str, None] = None,
+                 positive_confidence: float = 0.66,
+                 null_response: str = "I don't know",
+                 null_confidence: float = 0.33
+                 ) -> None:
```
[the remark this answers is not in the document]
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: fixed

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+        """
+        # grasp user_ids from states batch.
+        # We expect that skill receives None or dict of state for each utterance.
+        # if state has user_id then skill uses it, otherwise it generates user_id and calls the
+        # user with this name in further.
+
+        # In this implementation we use current datetime for generating uniqe ids
+        output_states_batch = []
+        user_ids = []
+        for each_state in states_batch:
```
[the remark this answers is not in the document]
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: ok

File: tests/test_aiml_skill.py [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+                                  "Learn my pants are Red",
+                                  "LET DISCUSS MOVIES",
+                                  "Comedy movies are nice to watch",
+                                  "I LIKE WATCHING COMEDY!",
+                                  "Ok, goodbye"
+        ]
+
+        history_of_responses = []
+        for each_utt in user_messages_sequence:
+            print(f"User says: {each_utt}")
```
[the remark this answers is not in the document]
Comment (acriptis) [reply tests/test_aiml_skill.py]: set logger responsibility:
```
            log.info(f"User says: {each_utt}")
            ...
            log.info(f" Bot says: {responses_batch[0]}")
```

File: deeppavlov/skills/aiml_skill/aiml_skill.py (hunk 18 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: add empty line before ```log = getLogger(__name__)```

File: tests/test_aiml_skill.py [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+            'null_response': "I don't know what to answer you",
+            'positive_confidence': 0.66
+        }
+        if path_to_aiml_scripts:
+            aiml_skill_config['path_to_aiml_scripts'] = str(path_to_aiml_scripts)
+
+        aiml_skill = AIMLSkill(**aiml_skill_config)
+        self.agent = DefaultAgent([aiml_skill], skills_selector=HighestConfidenceSelector())
+
+    def test_simple_reaction(self):
```
[the remark this answers is not in the document]
Comment (acriptis) [reply tests/test_aiml_skill.py]: May be I can include this into `test_quick_start.py`, but I need documentation about this script or consultation how to check that at particular iteration of test (first and fifth) component returns particular responses for particular user (`assert "Well, hello!" in history_of_responses[0][0]`)

And if this test may be implemented, I find that such tests are hard to read.

I can add to `test_quick_start` simple test that skill initializes correctly.

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+        """
+        Here you put user id generative logic if you want to implement it in the skill.
+
+        Although it is better to delegate user_id generation to Agent Layer
+        Returns: int
+
+        """
+        return uuid.uuid1()
+
+    def __call__(self, utterances_batch: list, history_batch: list, states_batch: list) -> Tuple[list, list, list]:
```
[the remark this answers is not in the document]
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Fixed

File: deeppavlov/skills/aiml_skill/aiml_skill.py (hunk 19 above)
Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: Really there is no requirements for file structure of AIML scripts. It is just a bunch of independent scripts which can be loaded to Kernel independently (but developer can inject dependencies into their scripts).

Examples of files are taken from here:
https://github.com/pandorabots/Free-AIML
(exactly I've taken only two scripts: `basic_chat.aiml`, `comedian.aiml`) and placed it into aiml_scripts folder.

This is described here:
https://github.com/deepmipt/DeepPavlov/blob/aiml_skill/deeppavlov/skills/aiml_skill/README.md
and here:
https://github.com/deepmipt/DeepPavlov/blob/aiml_skill/docs/skills/aiml_skill.rst

File: tests/test_aiml_skill.py [hunk 25]
```diff
@@ -0,0 +1,40 @@
+from pathlib import Path
+from logging import getLogger
+import pytest
+from deeppavlov.agents.default_agent.default_agent import DefaultAgent
+from deeppavlov.agents.processors.highest_confidence_selector import HighestConfidenceSelector
+from deeppavlov import configs, build_model
+from tests.test_quick_start import install_config
```
Comment (yoptar) [inline tests/test_aiml_skill.py]:
```suggestion
from .test_quick_start import install_config
```

File: tests/test_aiml_skill.py [hunk 26]
```diff
@@ -0,0 +1,40 @@
+from pathlib import Path
+from logging import getLogger
+import pytest
+from deeppavlov.agents.default_agent.default_agent import DefaultAgent
```
Comment (yoptar) [inline tests/test_aiml_skill.py]:
```suggestion

from deeppavlov.agents.default_agent.default_agent import DefaultAgent
```

File: tests/test_aiml_skill.py [hunk 27]
```diff
@@ -0,0 +1,40 @@
+from pathlib import Path
+from logging import getLogger
+import pytest
```
Comment (yoptar) [inline tests/test_aiml_skill.py]:
```suggestion

import pytest
```

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 28]
```diff
@@ -0,0 +1,139 @@
+import uuid
+from pathlib import Path
+from typing import Union
+from typing import Tuple, Optional, List
+from logging import getLogger
+import aiml
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]:
```suggestion

import aiml
```

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+from deeppavlov.core.common.registry import register
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]:
```suggestion

from deeppavlov.core.common.registry import register
```

File: tests/test_aiml_skill.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (my-master) [reply tests/test_aiml_skill.py]: To support the project coding style, I would suggest smth like this:

```
 logger = logging.getLogger("aiml").setLevel(logging.WARNING)
```

File: deeppavlov/skills/aiml_skill/aiml_skill.py [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+        # learn kernel to all aimls in directory tree:
+        all_files = sorted(self.path_to_aiml_scripts.rglob('*.*'))
+        learned_files = []
+        for each_file_path in all_files:
+            if each_file_path.suffix in ['.aiml', '.xml']:
+                # learn the script file
+                self.kernel.learn(str(each_file_path))
+                learned_files.append(each_file_path)
+        if not learned_files:
+            log.error(f"No .aiml or .xml files found for AIML Kernel in directory {self.path_to_aiml_scripts}")
```
Comment (yoptar) [inline deeppavlov/skills/aiml_skill/aiml_skill.py]: You should `raise` if it's an error.
Otherwise you can `log.warning()` instead of `log.error()`

Comment (acriptis) [reply deeppavlov/skills/aiml_skill/aiml_skill.py]: fixed

**Pull request merged**